### PR TITLE
Feature/refactoring

### PR DESCRIPTION
**Pull request by ikopylov**

### CODE REVIEW

File: BobClient/Qoollo.BobClient/BobNodeClient.cs [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+            }
+        }
+        /// <summary>
+        /// Explicitly opens connection to the Bob node
+        /// </summary>
+        /// <returns>Task to await</returns>
+        /// <exception cref="BobOperationException">Connection was not opened</exception>
+        public Task OpenAsync()
+        {
+            return OpenAsync(Timeout.InfiniteTimeSpan);
```
Comment (agend) [inline BobClient/Qoollo.BobClient/BobNodeClient.cs]: Infinite is too long

File: BobClient/BobClient.ConsoleApp/Program.cs [hunk 2]
```diff
@@ -7,23 +9,47 @@ namespace BobClient.ConsoleApp
 {
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/BobClient.ConsoleApp/Program.cs]: 👍

File: BobClient/Qoollo.BobClient/BobClusterClient.cs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+        {
+            if (timeout < TimeSpan.Zero && timeout != Timeout.InfiniteTimeSpan)
+                throw new ArgumentOutOfRangeException(nameof(timeout));
+
+            List<Exception> exceptions = new List<Exception>();
+            for (int i = 0; i < _clients.Length; i++)
+            {
+                try
+                {
+                    await _clients[i].OpenAsync(timeout);
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobClusterClient.cs]: GRPC can lazily open connection on the first call to the node, so this definitely can be simplified to just throwing on the first failure

File: BobClient/Qoollo.BobClient/BobClusterClient.cs [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+
+        /// <summary>
+        /// Explicitly opens connection to every Bob node in cluster
+        /// </summary>
+        /// <param name="timeout">Timeout</param>
+        /// <returns>Task to await</returns>
+        /// <exception cref="AggregateException">Aggregated exceptions from every node</exception>
+        public async Task OpenAsync(TimeSpan timeout)
+        {
+            if (timeout < TimeSpan.Zero && timeout != Timeout.InfiniteTimeSpan)
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobClusterClient.cs]: Yes, it is correct. `InfiniteTimeSpan` is equal to `-1` millisecond, all other negative values are incorrect

File: BobClient/Qoollo.BobClient/BobClusterClient.cs [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                throw new AggregateException(exceptions);
+        }
+        /// <summary>
+        /// Explicitly opens connection to every Bob node in cluster
+        /// </summary>
+        /// <returns>Task to await</returns>
+        /// <exception cref="AggregateException">Aggregated exceptions from every node</exception>
+        public Task OpenAsync()
+        {
+            return OpenAsync(Timeout.InfiniteTimeSpan);
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobClusterClient.cs]: 👍

File: BobClient/Qoollo.BobClient/BobNodeClient.cs (hunk 1 above)
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobNodeClient.cs]: 👍

File: BobClient/Qoollo.BobClient/BobClusterClient.cs [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+        /// <param name="timeout">Timeout</param>
+        /// <returns>Task to await</returns>
+        /// <exception cref="AggregateException">Aggregated exceptions from every node</exception>
+        public async Task OpenAsync(TimeSpan timeout)
+        {
+            if (timeout < TimeSpan.Zero && timeout != Timeout.InfiniteTimeSpan)
+                throw new ArgumentOutOfRangeException(nameof(timeout));
+
+            List<Exception> exceptions = new List<Exception>();
+            for (int i = 0; i < _clients.Length; i++)
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobClusterClient.cs]: This was simplified due to other changes, so I don't see the benefit of creating a separate function to perform operations on every node

File: BobClient/Qoollo.BobClient/NodeAddress.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    /// <summary>
+    /// Single node describtion
+    /// </summary>
+    public class NodeAddress
+    {
+        /// <summary>
+        /// Node constructor
+        /// </summary>
+        /// <param name="address">Node address. Format like host:port </param>
+        public NodeAddress(string address)
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/NodeAddress.cs]: This string goes directly to `Grpc.Core.Channel` constructor, which supports several different address formats. With explicit validation I can reject some formats that are currently supported by GRPC or will eventually become supported in the future.

File: BobClient/Qoollo.BobClient/BobNodeClient.cs [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+                case Grpc.Core.ChannelState.Connecting:
+                    return BobNodeClientState.Connecting;
+                case Grpc.Core.ChannelState.Ready:
+                    return BobNodeClientState.Ready;
+                case Grpc.Core.ChannelState.TransientFailure:
+                    return BobNodeClientState.TransientFailure;
+                case Grpc.Core.ChannelState.Shutdown:
+                    return BobNodeClientState.Shutdown;
+                default:
+                    throw new Exception($"Unexpected internal channel state: {state}");
```
[the remark this answers is not in the document]
Comment (ikopylov) [reply BobClient/Qoollo.BobClient/BobNodeClient.cs]: Replaced with `ArgumentException`

**Pull request merged**